Implement Authentication

tarzzz opened the pull request:
@chriddyp I have implemented a barebones version of how authorization will take place in DB Connector. 

It would be great if you could take a preliminary look and let me know if I am going in the right direction.. 😸 

Meanwhile, I would be working on implementing the actual logic and the process.

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,37 @@
+            // 2. Set username + generate another access-token, save it in local storage.
+            // Allow request to go through.
+        }
+
```
```
get('/v2/users/current', {'header': oauthtoken'}).then(apiResponse => {
    if(apiResponse.status === 200) {
        res.header.set_cookie = 'plotly-auth: token'
    }
    next();
})
```

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,37 @@
+        // TODO: Add exceptions for requests allowed without authentication
+        // i.e: homepage and /oauth/callback.
+
+        if request.authorization.scheme !== 'bearer' {
```
```
if (req.cookie['5-min-access-token'] === getAccessToken())
```

tarzzz commented on test/integration_test.js:
```diff
@@ -99,3 +99,78 @@ function clearConnectorFolder() {
+chromedriver.start(); // on port 9515
+process.on('exit', chromedriver.stop);
+
+describe('plotly database connector', function Spec() {
```
@chriddyp or @jackparmer it seems we got rid of this integration test in #181 merge.. Wanted to make sure that was intended, since it is the only integration test we had.

n-riesco replied:
It's still there: https://github.com/plotly/plotly-database-connector/blob/54c92ddcf83e1f0e5bde80aa0529b88c23513ebb/test/integration_test.js#L20-L25

tarzzz commented on test/integration_test.js:
```diff
@@ -99,3 +104,78 @@ function clearConnectorFolder() {
+
+    before(initialize);
+
+    it('should work', async () => {
```
@n-riesco Sorry, I meant this test. It was there before (I did not write it)

n-riesco replied:
I changed the description, but the test is the same.

tarzzz replied:
_sigh_, yeah.. I noticed it now. All good !!

n-riesco replied:
I guess you wnat to remove this:

tarzzz commented:
@chriddyp This is ready for a review.. 

There is one small tweak needed for electron apps, to automatically redirect from login page to connections page. Right now it relies on checking cookies, which is not available in electron.. but I can handle it as a separate PR.

tarzzz commented:
Update --  Added the electron tweak as well. It allows user to access electron app when not-logged in, but allows option to log-in as well:

![image](https://user-images.githubusercontent.com/1546532/30933434-8e1b7732-a398-11e7-92ac-1c8c356615df.png)

jackparmer commented:
Fantastic. So to confirm, you can use Falcom without logging in? People hate being forced to sign up for a platform when it's not necessary (ie they just want to query SQL and don't care about plotly), so I think this would be awesome.

tarzzz commented:
> So to confirm, you can use Falcom without logging in? People hate being forced to sign up for a platform when it's not necessary (ie they just want to query SQL and don't care about plotly), so I think this would be awesome.

For electron app --> Yes.

For headless/server version, we provide a configuration called `AUTH_ENABLED` which allows users/admins to switch off authentication as necessary (it is enabled by default)..

tarzzz commented:
on server when `AUTH_ENABLED=true`  :

![server-auth](https://user-images.githubusercontent.com/1546532/30936648-882bffcc-a3a2-11e7-88bf-81f29f6b17d9.gif)

tarzzz commented:
@chriddyp getting error on Mac Build:
![image](https://user-images.githubusercontent.com/1546532/30938743-174a3740-a3a9-11e7-91b5-d86a17a42d31.png)

Are there any specific steps we need to follow?

I am running:

```
$ yarn run package
$ ./zipItAll.sh
```

I have tried removing node_modules and re-installing + rebuilding everything as well..

chriddyp commented:
@tarzzz - `yarn run package` looks like the correct command. I'm not sure why this error is cropping up all of a sudden

tarzzz commented:
Headless build: https://plotly-connector-pr-180.herokuapp.com/

Still looking into the mac build issue..

n-riesco commented:
@tarzzz It shouldn't happen.
`source-map-support` is listed as a package dependency.
Could you check there is `source-map-support` in `release/[...]/Resources/app/node_modules`?
If it doesn't have `source-map-support`, wipe the `release` folder, make sure you've got enough hard drive space (> 600 MB) and rerun `yarn run package`.

tarzzz commented:
@n-riesco Even after trying `yarn run package` multiple times.. the `release/[...]/Resources/app/node_modules` is empty. There seems to be an issue with the package builder?

tarzzz commented:
```
$ release/Mac-64bit/Plotly Database Connector-darwin-x64/Plotly Database Connector.app/Contents/Resources/app/node_modules pwd
release/Mac-64bit/Plotly Database Connector-darwin-x64/Plotly Database Connector.app/Contents/Resources/app/node_modules
$ release/Mac-64bit/Plotly Database Connector-darwin-x64/Plotly Database Connector.app/Contents/Resources/app/node_modules ls
$ release/Mac-64bit/Plotly Database Connector-darwin-x64/Plotly Database Connector.app/Contents/Resources/app/node_modules
```

n-riesco commented:
@tarzzz Let's talk about this on Slack when you're around.

chriddyp commented on app/components/Configuration.react.js:
```diff
@@ -21,10 +22,19 @@ class Configuration extends Component {
+        this.logOut = this.logOut.bind(this);
+
+        if(window.require) {
+            window.require('electron').ipcRenderer.once('username',
```
is `if (window.require)` a check to see if we're in electron vs web browser? if so, could we move it into a re-usable `utils.js` file? I believe there is already something there called like `isElectron`

tarzzz replied:
`isElectron` is a node-module: https://www.npmjs.com/package/is-electron
but it is not working for us as expected (not sure why). So instead I am checking according to the context required ..

chriddyp replied:
got it. then can we remove `isElectron` and redefine it in `utils.js` like:
```
const isElectron = () => Boolean(window.require)
```

chriddyp commented on app/components/Configuration.react.js:
```diff
@@ -35,6 +45,22 @@ class Configuration extends Component {
+
+      // Delete all the cookies and reset user state. This does not kill
+      // any running connections, but user will not be able to access them
+      // without logging in again.
```
🐮  the convention for comments in this codebase is 
```
/*
 *
 *
 *
 */
```

chriddyp commented on app/components/Configuration.react.js:
```diff
@@ -35,6 +45,22 @@ class Configuration extends Component {
+      this.setState({ username: ''});
+
+      // reload page when running in browser:
+      if (window) {
```
🐮  could we use the same `utils.isElectron` function here?

chriddyp commented on app/components/Configuration.react.js:
```diff
@@ -21,10 +22,19 @@ class Configuration extends Component {
+            window.require('electron').ipcRenderer.once('username',
+                (event, message) => {
+                    this.setState({username: message});
+                });
```
could we add some comments here explaining what this is doing? Something along the lines of:
> If this is an electron app, then we don't have access to cookies and we don't require authentication in the API.
> In the browser setting, the username is set with a cookie but in electron this is set using electron's ipcRenderer.

chriddyp commented on app/components/Configuration.react.js:
```diff
@@ -43,9 +69,21 @@ class Configuration extends Component {
+                            :
+                            <div>
+                                <Link to="/login" >Log In</Link>
+                            </div>
```
🐮  this indentation looks off

chriddyp commented on app/components/Login.react.js:
```diff
@@ -52,25 +54,62 @@ class Login extends Component {
+                          statusMessage: 'Saving user information...'
+                      });
+                      that.saveDomainToSettings().then(res => {
+                          if ( res.status === 200 ) {
```
🐮  no spaces inside `(`

chriddyp commented on app/components/Login.react.js:
```diff
@@ -52,25 +54,62 @@ class Login extends Component {
+                          } else {
+                              that.setState(
+                                  status: 'failure',
+                                  statusMessage: 'An error occured. Please try again.'
```
🐮  spelling: `occurred`

chriddyp replied:
it would be helpful to display the error message from `res.text()` too, especially when we're helping customers out

chriddyp replied:
See https://github.com/plotly/plotly-database-connector/pull/180/files#diff-fd9a17590082224a502b26de40234d29L103 for an example of this error handling flow

chriddyp commented on app/components/Login.react.js:
```diff
@@ -52,25 +54,62 @@ class Login extends Component {
+            },
+            body: JSON.stringify({
+                PLOTLY_API_DOMAIN, PLOTLY_API_SSL_ENABLED
+            })
```
why is this needed? for on-prem, these settings will get set in run-time by the docker variables. for users running it themselves, they'll just use the default public cloud settings or modify the settings.yml file directly

chriddyp replied:
ah, i see that this was done below. Hm... was there a good reason for it? I can't think of any right now.

tarzzz replied:
I was not sure, so I kept it instead of deleting it. I am okay with removing it though.. 😸

tarzzz replied:
As discussed offline, we are leaving it as it is for now ..

chriddyp commented on backend/main.development.js:
```diff
@@ -32,6 +32,10 @@ app.on('ready', () => {
 
     const {httpServer, httpsServer} = server;
+
+    httpServer.electronWindow = mainWindow;
```
a comment here would be nice

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+var errors = require('restify-errors');
```
`const`? also, does `import errors from 'restify-errors'`  work?

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+let win = '';
+
+export function PlotlyOAuth(electronWindow) {
+    win = electronWindow;
```
🐮  can we remove this assignment? if it wasn't passed in, then it'll just be undefined and so the `if (electronWindow)` check should still work as intended

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+          return (next());
+        }
+        // Auth is disabled for certain urls:
+        if ( ESCAPED_ROUTES.some(path.match.bind(path)) ) {
```
🐮  rm spaces in `()`

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+    win = electronWindow;
+
+    function isAuthorized(req, res, next) {
+        let path = req.href();
```
can this be `const`? it doesn't look like it's being reassigned.

we really gotta get `eslint` in our tests.

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+        if (!accessTokenIsValid(req.cookies['db-connector-auth-token'])) {
+
+            if (!req.cookies['plotly-auth-token']) {
+              return (next(new errors.InvalidCredentialsError('Please login to access this page.')));
```
this error handling is different than how we do error handling in the rest of the routes. I'd prefer that we'd use one error handling library. since we aren't passing this request forward to anything else, can this just be:
```
res.json(401, {error: {message: 'Please login to access this page'}});
return;
```

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+
+        }
+
+        return (next());
```
shouldn't this be wrapped in an `else {}`? The fetch in the line above is asynchronous, so it may not enter its `.then` before this `next()` has been called.

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+              return (next(new errors.InvalidCredentialsError('Please login to access this page.')));
+            }
+
+            const plotly_access_token = req.cookies['plotly-auth-token'];
```
🐮  `plotlyAuthToken`

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+            })
+            .then(userRes => userRes.json().then(userMeta => {
+
+              if ( userRes.status !== 200 ) {
```
🐮  :zap: space around parentheses

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+              }
+              else {
+                  const db_connector_access_token = generateAndSaveAccessToken();
+                  res.setCookie('db-connector-auth-token', db_connector_access_token, {'maxAge': 300, 'path': '/'});
```
where is the check to see if this user _is allowed_ to view this route? shouldn't there be a check like:
```
if (contains(userMeta.username, getSetting('ALLOWED_USERS')) {
    res.setCookie(...);
} else {
   res.json(401, {error: {message: `User ${userMeta.username} is not allowed to view this app`}})
}
```

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+              }
+              else {
+                  const db_connector_access_token = generateAndSaveAccessToken();
+                  res.setCookie('db-connector-auth-token', db_connector_access_token, {'maxAge': 300, 'path': '/'});
```
For `path` - on on-prem, the connector will be under the url `plotly.acme.com/database-connector` or something. i believe that we should set `path` to that url prefix for on-prem.

tarzzz replied:
setting `PATH` as `/` will ensure that it will cover both the scenarios (i.e. `plotly.acme.com/database-connector` for on-prem, and any other url for standalone server-apps) .. 

> Path=<path-value> Optional
Indicates a URL path that must exist in the requested resource before sending the Cookie header. The %x2F ("/") character is interpreted as a directory separator and sub directories will be matched as well (e.g. path=/docs, "/docs", "/docs/Web/", or "/docs/Web/HTTP" will all be matched).

Let me know your thoughts on the same..

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+              }
+              else {
+                  const db_connector_access_token = generateAndSaveAccessToken();
+                  res.setCookie('db-connector-auth-token', db_connector_access_token, {'maxAge': 300, 'path': '/'});
```
we should also add a `secure` flag appropriately

tarzzz replied:
Adding a `secure` flag will make cookies not being set (or sent back to server) when we are using headless version _without_ SSL / HTTPS Enabled .. 

 It may cause issues with some on-prem installations..

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,68 @@
+  new RegExp('^/$'),
+  new RegExp('^/login$'),
+  new RegExp('^/static/'),
+  new RegExp('^/oauth2')
```
missing `$`?

chriddyp commented on backend/routes.js:
```diff
@@ -258,6 +267,21 @@ export default class Servers {
+                    res.setCookie('plotly-auth-token', access_token, {'path': '/'});
+
+                    const db_connector_access_token = generateAndSaveAccessToken();
+                    res.setCookie('db-connector-auth-token', db_connector_access_token, {'maxAge': 300, 'path': '/'});
```
same comments as in https://github.com/plotly/plotly-database-connector/pull/180/files#r141751696

chriddyp commented on backend/routes.js:
```diff
@@ -258,6 +267,21 @@ export default class Servers {
                         status = 201;
                     }
                     saveSetting('USERS', existingUsers);
+
```
similiarly to https://github.com/plotly/plotly-database-connector/pull/180/files#r141751154, it looks like we're missing the check that verifies that this user is allowed to view this app

chriddyp commented on backend/routes.js:
```diff
@@ -258,6 +267,21 @@ export default class Servers {
+
+                    if (that.electronWindow) {
+
+                        that.electronWindow.loadURL(`${protocol}://${domain}:${port}/`);
```
:dolls: a comment here explaining that this redirect is done on the front-end for the browser version of this app might be helpful

chriddyp commented on backend/utils/authUtils.js:
```diff
@@ -0,0 +1,8 @@
+const uuidv4 = require('uuid/v4');
```
does `import` work here?

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,58 +216,93 @@ describe('Routes - ', () => {
-            assert.equal(json.https, '');
-            done();
-        }))
+    it('allows access to connections page without logging in', function(done) {
```
i don't quite understand this - is the "connections" page the page with all of the database credentials? shouldn't that page be authenticated?

tarzzz replied:
The url doesn't requires authentication, since it is the home page. But a user cannot see any connections without logging in (Since other requests to `/connections` etc are not allowed for un-authenticated users:


![image](https://user-images.githubusercontent.com/1546532/31029746-edde2e74-a520-11e7-9abd-362068c7b92c.png)

tarzzz replied:
We can probably have a better UI for the homepage for unauthenticated users though..

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,58 +216,93 @@ describe('Routes - ', () => {
+            }).catch(done);
+        });
+
+    });
```
could we add some tests that test the behaviour of:
- making a request after the 5 min expiry token expires makes another request to plotly
- making a request within the 5 min expiry doesn't make a request to plotly
- if a user is revoked within 5 minutes then they can still access the connector but after the token expires they are denied access
- loop through all of the `GET` endpoints and verify that 401s are returned. i believe that you can access the list of routes dynamically with `server.router.mounts`
- requests to `/`, `/login/`, `/static/`, and `/oauth2/` return in 200s

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -262,6 +315,9 @@ describe('Routes - ', () => {
 
+    // This tests sets the cookies required for user authentication, so
+    // all the tests requiring user authentication should be put placed
+    // after this one.
```
🐮  comment style with `/*`

chriddyp replied:
could we move this into a `beforeEach`? otherwise, it'll be impossible to test an individual test

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -312,6 +368,51 @@ describe('Routes - ', () => {
+            assert.equal(res.status, 200);
+            assert.equal(json.http, 'http://localhost:9494');
+            assert.equal(json.https, '');
+            done();
```
could we add `assert.equal(R.keys(json), ['http', 'https'])`. we want to make sure that secret settings aren't accidently leaked

chriddyp replied:
or a `deepEqual` like the settings test below

tarzzz commented:
Update: Addressed the comments and fixed existing tests.. 

I am working on covering these test-cases:

>  - making a request after the 5 min expiry token expires makes another request to plotly
>  - making a request within the 5 min expiry doesn't make a request to plotly
 > - if a user is revoked within 5 minutes then they can still access the connector but after the token expires they are denied access
>  - loop through all of the GET endpoints and verify that 401s are returned. i believe that you can 
     access the list of routes dynamically with server.router.mounts
 > - requests to /, /login/, /static/, and /oauth2/ return in 200s

The last one has already been covered.. so working on others..

tarzzz commented:
@chriddyp Addressed the comments and added some tests related to accessTokens (as mentioned above).. Please let me know if it looks good to you !! 😸

chriddyp commented on app/components/Login.react.js:
```diff
@@ -52,25 +54,61 @@ class Login extends Component {
+    componentDidMount() {
+        const {serverType} = this.state;
+        const interval = setInterval(() => {
+              usernameLogged = cookie.load('db-connector-user');
```
could we add a comment here explaining why we're only looking at cookies and not the ipc state? in other words, why this block of code doesn't pertain to the electron environment

chriddyp commented on app/components/Login.react.js:
```diff
@@ -52,25 +54,61 @@ class Login extends Component {
+                          } else {
+                              this.setState({
+                                  status: 'failure',
+                                  statusMessage: res.text()
```
`res.text()` is a promise, so you need a .`then(text => this.setState({statusMessage: text, status: 'failure'})` to get the actual text

chriddyp commented on app/components/Login.react.js:
```diff
@@ -52,25 +54,61 @@ class Login extends Component {
+                                  statusMessage: res.text()
+                              });
+                          }
+                      });
```
it looks like we're missing a `.catch()`

chriddyp commented on backend/utils/authUtils.js:
```diff
@@ -0,0 +1,28 @@
+
+export function isElectron() {
+
+    if (process.versions && !!process.versions.electron) {
```
🐮  instead of `!!`, I prefer something more explicit like `R.has('electron', process.versions)` (in general, I try to use `ramda` wherever possible - it's like our "standard library")

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,87 @@
+                      res.setCookie('db-connector-auth-token',
+                                    dbConnectorAccessToken, {
+                                        'maxAge': getAccessTokenExpiry(),
+                                        'path': '/'
```
I think we should make `secure` a setting like `SSL_ENABLED` - this is similar to how the dash app manager works: https://github.com/plotly/dokku-integration/blob/master/manager/backend/server.py#L99-L105
And I still think that the more specific the path, the better. This will allow users to run this behind a proxy and reduce the thread of a cookie leaking through an XSS in their main app code. Also similar to how the app manager works: https://github.com/plotly/dokku-integration/blob/master/manager/backend/server.py#L103

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -0,0 +1,87 @@
+
+                      // Remove any existing credentials and return error
+                      res.clearCookie('db-connector-auth-token');
+                      res.clearCookie('plotly-auth-token');
```
clear `db-connector-user` as well?

chriddyp commented on backend/routes.js:
```diff
@@ -258,7 +267,37 @@ export default class Servers {
                         status = 201;
                     }
                     saveSetting('USERS', existingUsers);
-                    res.json(status, {});
```
Does it make sense to move the `saveSetting('USERS', existingUsers)` to after the user has been validated?

tarzzz replied:
Yep, make sense to me.. Updated !!

chriddyp commented on backend/routes.js:
```diff
@@ -258,7 +267,37 @@ export default class Servers {
+                        const db_connector_access_token = generateAndSaveAccessToken();
+                        res.setCookie('db-connector-auth-token',
+                                      db_connector_access_token, {
+                                          'maxAge': getAccessTokenExpiry(),
```
should this just be `db_connector_acccess_token`?

tarzzz replied:
Sorry, dont get it.. you mean the cookie-name should be `db-connector-access-token` instead of `db-connector-auth-token` ..?

chriddyp replied:
I actually don't know what I was referring to.

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -85,10 +89,26 @@ function DELETE(path) {
 let connectionId;
 
+let isElectronStub = sinon.stub(utils, 'isElectron');
+isElectronStub.returns(false);
```
nice!

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -85,10 +89,26 @@ function DELETE(path) {
 let servers;
 let connectionId;
 
+let isElectronStub = sinon.stub(utils, 'isElectron');
```
🐮  `const`

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -85,10 +89,26 @@ function DELETE(path) {
+let isElectronStub = sinon.stub(utils, 'isElectron');
+isElectronStub.returns(false);
+
+let accessTokenExpiryStub = sinon.stub(utils, 'getAccessTokenExpiry');
```
🐮  `const`

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -170,6 +192,9 @@ describe('Routes - ', () => {
         });
 
+        // enable authentication:
+        saveSetting('AUTH_ENABLED', 'true');
```
I believe that booleans work here: `true`

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
-        }))
+    it('allows access to connections page without logging in', function(done) {
+        GET('')
+        .then(() => done())
```
note that calling `GET` isn't enough to show that it's allowed as `fetch` doesn't raise an error for non-200 level status codes. You'll have to explicitly do `GET('').then(res => {assertEqual(res.status, 200); done()})`

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+        POST('oauth2', {access_token: accessToken})
+        .then(res => {
+            // request should be allowed:
+            GET('settings').then(res => {
```
missing a `return` here for proper promise chaining

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+
+        }).catch(done);
+    });
+
```
tests look good! i think that the only missing behaviour is that "renewal works" - i.e. testing the lines below this: https://github.com/plotly/plotly-database-connector/pull/180/files#diff-2d3ed64023e43dc5da0d43e8e5df7040R49. 

I think that you can test this behaviour withs omething like:

```
// set access-token expiry of 3 sec:
expiry = 10;
accessTokenExpiryStub.returns(expiry);

POST('oauth2', {access_token: accessToken})
.then(res => {
    // request should be allowed:
    return GET('settings').then(res => {
      console.log('status1:' + res.status);
      assert.equal(res.status, 200);
    });
})
.then(() => {
    // should renew after expired and still allow access
    setTimeout(() => {
        GET('settings').then(res => {
            console.log('status2:' + res.status);
            assert.equal(res.status, 200);
            done();
        }).catch(done);
    }, (expiry + 3) * 1000)
}).catch(done);
```

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+                    done();
+                })).catch(done);
+            }, 2000);
+        });
```
missing a `.catch(done)` here

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+                assert.equal(res.status, 200);
+                done();
+            }).catch(done);
+        });
```
missing a `.catch(done)` here

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+                assert.equal(res.status, 200);
+                done();
+            }).catch(done);
+        });
```
missing a `.catch(done)` here

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+            .then(res => {
+                assert.equal(res.status, 200);
+                done();
+            }).catch(done);
```
if you `return` this `GET` then i believe that you can remove this `.catch(done)` (as long as you add it to the very end)

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+        .then(() => {
+
+            // Any requests should fail after revoking user access
+            GET('settings').then(res => {
```
missing a `return` here

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+              assert.equal(res.status, 200);
+            });
+        })
+        .then(saveSetting('ALLOWED_USERS', []))
```
this isn't asynchronous so you could move this into the next `.then` block

chriddyp replied:
also, this isn't a function, so this is going to be immediately invoked. the proper way to write this would be `.then(() => saveSetting('ALLOWED_USERS', []))` (but you can just move it into the next `.then` block

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -198,60 +223,192 @@ describe('Routes - ', () => {
+
+        // Login the user:
+        saveSetting('ALLOWED_USERS', [username]);
+        setCookies(done);
```
since this calls `done`, should this be at the very end of this `beforeEach` function? if this were synchronous, then the following lines wouldn't get called

chriddyp commented:
woohoo! 💃  once my comments are addressed. alternatively, I'm happy to review it again if you'd like.

tarzzz commented:
@chriddyp Addressed the comments. let me know if it, feel good to be merged.. 😸

chriddyp commented on backend/main.development.js:
```diff
@@ -16,7 +16,9 @@ const isTestRun = contains('--test-type=webdriver', process.argv.slice(2));
 
 const server = new Servers();
 Logger.log('Starting server', 2);
+server.isElectron = true;
```
👍

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -65,7 +66,8 @@ export function PlotlyOAuth() {
                                         'maxAge': getAccessTokenExpiry(),
-                                        'path': '/'
+                                        'path': getSetting('DB_CONNECTOR_URL'),
+                                        'secure': getSetting('PLOTLY_API_SSL_ENABLED')
```
I think that we need a separate setting for this. Something like `CONNECTOR_SSL_ENABLED` or even just `SSL_ENABLED`. 

If someone is running this on their own servers against plot.ly, then they may or may not have SSL enabled even though https://plot.ly is.

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -65,7 +66,8 @@ export function PlotlyOAuth() {
                                     dbConnectorAccessToken, {
                                         'maxAge': getAccessTokenExpiry(),
-                                        'path': '/'
+                                        'path': getSetting('DB_CONNECTOR_URL'),
```
Maybe just `WEB_BASE_PATHNAME` would be a simpler setting. I think that `DB_CONNECTOR` is a little bit redundant since the settings implicitly refer to itself anyway. And URL is usually the full URL (`https://domain/pathname`) and we just need the `pathname`

chriddyp commented on backend/routes.js:
```diff
@@ -266,20 +266,22 @@ export default class Servers {
 
                     if (contains(username, getSetting('ALLOWED_USERS'))) {
-                        res.setCookie('plotly-auth-token', access_token, {'path': '/'});
+                        res.setCookie('plotly-auth-token', access_token, {'path': getSetting('DB_CONNECTOR_URL')});
```
it looks like this is missing `secure`. We should probably keep this dry and make these options a constant like
```
COOKIE_OPTIONS = {
    'path': ...,
    'secure': '...
}
```

chriddyp commented on backend/routes.js:
```diff
@@ -257,8 +266,40 @@ export default class Servers {
+                    if (contains(username, getSetting('ALLOWED_USERS'))) {
+                        res.setCookie('plotly-auth-token', access_token, {'path': getSetting('DB_CONNECTOR_URL')});
+
+                        const db_connector_access_token = generateAndSaveAccessToken();
```
I just noticed that this the flow will _reset_ the access token. So, if 5 people log in at the same time, the access token will get reset after each person. 
I got around this in the dash-auth package by saving the expiry date of the access token and my `generateAndSaveAccessToken` looked at the date before generating:
```
generateAndSaveAccessToken() {
    if (getSetting('ACCESS_TOKEN_EXPIRATION') > now()) {
         const access_token = uuid4();
         saveSetting('ACCESS_TOKEN', access_token);
         saveSetting('ACCESS_TOKEN_EXPIRATION', now + 5 minutes);
         return access_token;
    } else {
        return getSetting('ACCESS_TOKEN');
    }
}
```

It's an uncommon case but maybe should tighten it up

chriddyp commented on backend/routes.js:
```diff
@@ -257,8 +266,40 @@ export default class Servers {
+                                          'path': getSetting('DB_CONNECTOR_URL'),
+                                          'secure': getSetting('PLOTLY_API_SSL_ENABLED')
+                                      });
+                        res.setCookie('db-connector-user', username, {'path': '/'});
```
let's make these cookie options a `CONSTANT` as in https://github.com/plotly/plotly-database-connector/pull/180/files/e7034e717d8602fa95158af34447b5a8d4e7f712..7204ec66554f53e5732be3d1da64fcd020c0323d#r143544790 so that we can be sure that they're consistent everywhere

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -269,9 +271,10 @@ describe('Authentication - ', () => {
-        });
+                return;
+            });
+            done();
```
This `done()` is outside of the `.then` that includes the `assert.equal(res.status, 200)` and so the `assert.equal` won't get called because the `done` will get called before it. So, should keep the `done()` in the `.then`

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -269,9 +271,10 @@ describe('Authentication - ', () => {
             GET('settings')
```
In order for errors inside the `.then` of this functions to bubble up to the `.catch` on very outside, you'll need to `return` this `GET` (I believe)

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -292,9 +295,10 @@ describe('Authentication - ', () => {
+                return;
+            });
+            done();
+        }).catch(done);
```
https://github.com/plotly/plotly-database-connector/pull/180/files/e7034e717d8602fa95158af34447b5a8d4e7f712..7204ec66554f53e5732be3d1da64fcd020c0323d#r143551641 and https://github.com/plotly/plotly-database-connector/pull/180/files/e7034e717d8602fa95158af34447b5a8d4e7f712..7204ec66554f53e5732be3d1da64fcd020c0323d#r143551795 apply here as well

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -319,10 +323,37 @@ describe('Authentication - ', () => {
+                }));
             }, 2000);
-        });
+            done();
```
https://github.com/plotly/plotly-database-connector/pull/180/files/e7034e717d8602fa95158af34447b5a8d4e7f712..7204ec66554f53e5732be3d1da64fcd020c0323d#r143551641 and https://github.com/plotly/plotly-database-connector/pull/180/files/e7034e717d8602fa95158af34447b5a8d4e7f712..7204ec66554f53e5732be3d1da64fcd020c0323d#r143551795 apply here as well

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -319,10 +323,37 @@ describe('Authentication - ', () => {
                                  'bad-domain.plot.ly bad-domain.plot.ly:443'));
-                    done();
-                })).catch(done);
+                    return;
```
since this `setTimeout` isn't a promise, errors inside it won't get bubble out to the outermost promise and `.catch`. So,  this `.catch(done)` needs to stay

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -319,10 +323,37 @@ describe('Authentication - ', () => {
+                    return;
+                }));
+            }, 3000);
+            done();
```
Similar comments in https://github.com/plotly/plotly-database-connector/pull/180/files/e7034e717d8602fa95158af34447b5a8d4e7f712..7204ec66554f53e5732be3d1da64fcd020c0323d#r143552194 - 
`setTimout` is asynchronous and it is not a promise, so in this case `done()` is going to be called before `setTimeout` is finished. `done()` needs to be where the `return;` statement is inside `setTimeout`

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -292,9 +295,10 @@ describe('Authentication - ', () => {
-                done();
-            }).catch(done);
-        });
+                return;
```
these returns aren't necessary

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -319,10 +323,37 @@ describe('Authentication - ', () => {
+                GET('settings')
+                .then(res => res.json().then(json => {
+                    assert.equal(res.status, 200);
+                    return;
```
(the `done()` should be here)

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -319,10 +323,37 @@ describe('Authentication - ', () => {
+                .then(res => res.json().then(json => {
+                    assert.equal(res.status, 200);
+                    return;
+                }));
```
a `.catch(done)` should be here

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -319,10 +323,37 @@ describe('Authentication - ', () => {
+                }));
+            }, 3000);
+            done();
+        }).catch(done);
```
this can stay 👍

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -319,10 +323,37 @@ describe('Authentication - ', () => {
+
+        POST('oauth2', {access_token: accessToken})
+        .then(res => {
+            GET('settings').then(res => res.json().then(json => {
```
missing a `return` here

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -319,10 +323,37 @@ describe('Authentication - ', () => {
+        .then(res => {
+            GET('settings').then(res => res.json().then(json => {
+                assert.equal(res.status, 200);
+                return;
```
this `return` can ⚡️

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -336,20 +367,20 @@ describe('Authentication - ', () => {
-              console.log('status1:' + res.status);
-              assert.equal(res.status, 200);
+                assert.equal(res.status, 200);
+                return;
```
I've created a little promise example for you to play around with so that it is crystal clear how this works: https://codepen.io/chriddyp/pen/boMNEz?editors=0011

tarzzz commented:
@chriddyp -- Added updates:

  - Updated settings names,
  - Added `ACCESS_TOKEN_EXPIRY` logic while re-generating token, as well as while authenticating,
  - Fixes promises related issue in tests. Thanks for such an informative codepen example.. 😸

There are some test failures (which are working fine locally for me), working on a fix.. but let me know if code looks good otherwise..

chriddyp commented on backend/constants.js:
```diff
@@ -0,0 +1,11 @@
+}
+
+export const ACCESS_TOKEN_COOKIE_OPTIONS = assoc(
+   'maxAge', getSetting('ACCESS_TOKEN_AGE'), COOKIE_OPTIONS);
```
Right now these will be evaluated on import which might not actually work in the test environment (where we change the settings on runtime). So, we might want to change to something like:
```
export function getCookieOptions() {
    return {
        secure: getSetting('SSL_ENABLED'),
        path: getSetting('WEB_BASE_PATHNAME')
    }
}
```

chriddyp commented on backend/plugins/authorization.js:
```diff
@@ -12,7 +13,9 @@ const ESCAPED_ROUTES = [
-    return getSetting('ACCESS_TOKEN') === access_token;
+    const currentTime = Date.now();
+    return (getSetting('ACCESS_TOKEN') === access_token &&
+            getSetting('ACCESS_TOKEN_EXPIRY') > currentTime);
```
We might need to `parseInt` this setting. I'm not sure.

chriddyp replied:
if it's coming from the `env`, then we `JSON.parse` it which will turn it into a number: https://github.com/plotly/plotly-database-connector/blob/master/backend/settings.js#L160

```
> typeof JSON.parse('4')
'number'
```

but i'm not sure what the type will be if it comes from YAML

chriddyp replied:
it's probably worth throwing in a quick test in https://github.com/plotly/plotly-database-connector/blob/master/test/backend/settings.spec.js for it:

```
it('Parses number settings as numbers', ()  => {
    const now = Date.now();
    saveSetting('ACCESS_TOKEN_EXPIRY', now);
    const nowSetting = getSetting('ACCESS_TOKEN_EXPIRY');
    expect(typeof nowSetting, 'number');
    expect(nowSetting, now);
    
    const anotherNow = now + 1;
    process.env.PLOTLY_CONNECTOR_ACCESS_TOKEN_EXPIRY = String(anotherNow);
    const anotherNowSetting = getSetting('ACCESS_TOKEN_EXPIRY');
    expect(typeof anotherNowSetting, 'number');
    expect(anotherNowSetting, anotherNow);
});
```

That way we can rest easy if ever update our YAML dep or otherwise modify `getSetting`

tarzzz replied:
Checked, and it seems it is `number` so we should be good w/o `parseInt`. Moreover if this behavior changed in future, ⬆️ these tests (added!) will let us know.. :)

chriddyp commented:
looking good! just a couple of more small comments

tarzzz commented:
woohoo.. merging..  🎉
